### ISSUE TITLE
Migrate from Portable Class Libraries to .NetStandard

### ISSUE THREAD

**Issue by seesharper** [dotnet/corefx#11429]
I am the author of the LightInject DI framework and I have a question regarding the .Net Platform Standard and how this is supposed to work with regards to different runtimes.

LightInject is basically cross compiled into 4 different binaries.

* Full Net 45
* Full Net 46
* .NetCore
* Portable Class Library, Profile 111

LightInject uses Reflection.Emit on all platforms except for in the PCL version where we use expression tree's to dynamically build code which works great.

I am fully aware of the restrictions that applies to AOT (Ahead of time) platforms and that Reflection.Emit is a "no-go" on platforms such as iOS and so on. 

What I don't understand is how the .Net Platform Standard manages these platform/runtime specific restrictions.

Taken from the Observations section
"If a library targets .NET Platform Standard version 1.3, it can only run on .NET Framework 4.6 or later, .NET Core, Universal Windows Platform 10 (UWP), and Mono/Xamarin platforms."

Okay, so as long as I compile for .NetStandard 1.3 it will be able to run on UWP? 

The thing is that you can't since there is no support for Reflection.Emit on UWP although having my assembly compiled for .NetStandard 1.3 suggests that it should work. 

What piece of the puzzle am I missing here?

**Comment by GuardRex** [comment]
I have an upcoming question in this area (specifically wrt the directive to use) for a new project I'm about to get into.

> The thing is that you can't since there is no support for Reflection.Emit on UWP although having my assembly compiled for .NetStandard 1.3 suggests that it should work.

The .NET Platform Standard version grants access to the indicated platforms. They don't guarantee that a particular class or method will be provided on a given platform just because it exists on other platforms at the same version of the .NET Standard. You get whatever each platform normally offers in the way of API surface area. <- @davidfowl will correct me if that language misses the mark.

Word on the street (@davidfowl) is that you get preprocessor defines for the framework TFM's automatically: https://github.com/aspnet/Announcements/issues/3#issuecomment-92196459

So ... and using @seesharper OP in the example ... would it be like this ...
```c#
#if UAP10_0,  // or NETCORE50 (I think)
    // Do something in place of Reflection.Emit
#else
    // Use Reflection.Emit
#endif
```
I also noticed that in some spots, MS devs are using `NETNATIVE`. [Example 1](https://github.com/dotnet/corefx/blob/d0dc5fc099946adc1035b34a8b1f6042eddb0c75/src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.PrivateKey.cs#L30-L38) and [Example 2](https://github.com/dotnet/corefx/blob/3a2d45cea19e91f1add0adaa53c50ba867ee440d/src/System.Security.Cryptography.Cng/src/System/Security/Cryptography/ECDsaCng.cs#L74-L78). Did they define that themselves, or is `NETNATIVE` an automatically-available one?

**Comment by seesharper** [comment]
If that is the case, it would be rather difficult to switch from a PCL library to a .NetStandard library as PCL libraries guaranties the surface area where as .Net Standards libraries does not?

I understand the intent, but only briefly understand how this actually works. PCL's are certainly not perfect, but at least understandable :)

For someone new to .Net and the whole story around TFM's, frameworks, runtimes and not to mention the "Bait and switch" technique , must be daunting to say the least :)

**Comment by GuardRex** [comment]
> must be daunting to say the least

Indeed, I had a good JabbR chat with @davidfowl about it over a year ago. He assured me that it was/is well explained and that lib authors would get it ... eventually, right? :smile:

Devs have found @davidfowl's analogy doc helpful ... See if that makes more sense than the cold, hard textual explanations: https://gist.github.com/davidfowl/8939f305567e1755412d6dc0b8baf1b7

**Comment by seesharper** [comment]
I remember @davidfowl tweeted about this some time ago, but it only confirms what I have been able to extract myself from the "cold hard textual explanations" :)  

The following interface perfectly explains what I don't understand. 

Types such as AssemblyBuilder (Reflection.Emit) are defined in INetStandard1.0 and hence inherited down all the way to UWP suggesting compatibility where there really is not. 

```
    interface IWindowsUniversalPlatform : INetStandard14
    {
        void GPS();
        void Xaml();
    }
```

One of the comments from the gist 
"Package Authors: Target the lowest netstandard version you can. You will run on all platforms that support that netstandard version or higher"

It seems to me that the concept is all about adding new features for each new .Net Platform Standard and not being able to restrict the surface area like PCL's do. 

The solution in the case for LightInject is to create a UAP moniker in the Nuget Package containing the PCL 111 binary which is contradictory since that is what is the .Net Standard is suppose to solve being open ended and all :) 

There is a good chance that I am just being plain stupid about this, but I have been working with .Net for 13 years and the .Net space has never been more confusing than now :)

**Comment by davidfowl** [comment]
Unfortunately, some of the features that masquerade as netstandard are not implemented on all platforms (that's just a reality of the system). It's a bait and switch approach where you can expose a NETStandard API but the implementation can only run on certain platforms. For these sorts of gaps it would be good to add a runtime API to detect the capabilities of the system when it can't be expressed solely by API surface e.g. https://github.com/dotnet/corefx/issues/10470

> There is a good chance that I am just being plain stupid about this, but I have been working with .Net for 13 years and the .Net space has never been more confusing than now :)

.NET runs on:
- Windows Desktop
- Windows Phone
- IPhone
- Android
- Playstation
- RaspberryPI
- IOT devices
- OSX
- Linux 

You get the point. The ecosystem is much more diverse and complicated than just windows so this is something you'll have to get accustomed to as a .NET developer. That said, we're doing the best we can to make the experience as smooth as possible.

/cc @ericstj @weshaggard

**Comment by seesharper** [comment]
@davidfowl It might be an idea to point this out in the .Net Platform Standard document so that library developers don't make false assumptions about where their code will actually run. 
As the document stands today it might lead the developer down the wrong path.

All this can become even more confusing when you suddenly encounter things that works on platforms where it should not be possible. LambdaExpression.Compile is a good example of this. 
When an expression is turned into IL on platforms supporting dynamic code generation, the expression tree is interpreted at runtime on AOT platforms such as iOS. It's slow, but it works.

When I first read the .Net Platform Standard I thought that , hey, there is no reason they could not do this even for the ILGenerator. 

When I think about it you could actually support System.Reflection.Emit.ILGeneration (DynamicMethod) on UAP since expression trees maps nicely to IL code, but that is probably another discussion :)

**Comment by davidfowl** [comment]
I 1000% agree 😄 

> When an expression is turned into IL on platforms supporting dynamic code generation, the expression tree is interpreted at runtime on AOT platforms such as iOS. It's slow, but it works.

Happens all of the time.

> When I think about it you could actually support System.Reflection.Emit.ILGeneration (DynamicMethod) on UAP since expression trees maps nicely to IL code, but that is probably another discussion :)

I'd actually like to bring this up as a general discussion with DI authors. AOT and the current DI systems are at odds (most runtime code generation in general actually). But yes a conversation for another day 😄

**Comment by seesharper** [comment]
When I first started out with the LightInject DI framework years ago, I decided to go with DynamicMethod and IL generation. When WinRT came along which did not support dynamic code generation through DynamicMethod, I created my own ILGenerator (#ifdef) that created the corresponding expression tree behind the scenes. In that way I could use IL Generation on all platforms. Same API surface :)

It is an interesting idea actually supporting lightweight IL generation using an adapter over expression trees. Suddenly a huge amount of libraries that relies on this would suddenly just work.
AssemblyBuilder, TypeBuilder and all that stuff is rarely needed unless you are doing a proxy library :)


As you said, another day :)

**Comment by stalek71** [comment]
The issue is with whole bunch of libs relying on dynamic proxies of course (like moq and nsubstitute to name a few only :) ) I heard MS created moq.4 some time ago for internal usage, but I don't now if it succeeded or was just kind of experiment only (and worked with UWP for Win10 at all).  

The only convenient workaround to the problem (with unit testing and usage of DI there) I know at the moment, is to use something like ProjectLinker extension spying events in VS and linking files between uwp project and "classic" .net unit tests project (so when you create ViewModel file in uwp it appears as linked file in unit test project so you can test it using moq and automock for DI dependency resolving on the fly).
It doesn't resolve core issue with DI on .NetCore (and AOT) but is quite good workaround at the moment for issue with unit testing and ensuring proper code quality on UWP at all.
I tested also DryIoC with UWP and it worked fine on UWP/.NetCore so, as @seesharper mentioned already, it' possible to achieve such goal like availability of quite fast DI containers working without using Emit method if only developers of DI are aware what is possible on what platform.

btw: there is such feature requested on [users voice](https://wpdev.uservoice.com/forums/110705-universal-windows-platform/suggestions/8280606-windows-10-uwp-should-support-reflection-emit-for), although I'm skeptic if it could be implemented because of restrictions on iOS for instance :( 

It was just as "bottom line" comment :) ...

**Comment by weshaggard** [comment]
I agree that having our bait-and-switch libraries like Ref.Emit have a netstandard ref and only be supported on some platforms is a terrible experience for library developers without guardrails. @terrajobst @ericstj do you know what our current plans are around supporting guardrails for these type of scenarios?  We definitely need to improve our developer experience here or rethink how we are packaging some of our libraries.

**Comment by gulbanana** [comment]
linking issue #9039 as this one is addressing the same questions

**Comment by tarekgh** [comment]
@terrajobst @ericstj any input regarding this one?

**Comment by ericstj** [comment]
All the libraries that folks are talking about here were designed to use NuGet's guardrails.  In that way they weren't pretending to be PCLs any more than any other fat package.  We were counting on tooling that would tell folks when they were using the packages, where those packages would run.  This was the `guardrails` feature @weshaggard mentioned and it was meant to be a requirement for folks who cared about *ensuring* their libraries would run everywhere.  We made a guarantee for everything in the NETStandard.Library closure that it'd be supported everywhere but for stuff above that you needed to use guardrails to be safe.  That tooling turned out to be flaky at best, and it wasn't consistently used (only PCL projects used it, none of the .NET Core projects did), so it was abandoned.  We're left in a place where we no longer have a way to represent "partial" portability at build time.

There still is no sufficient replacement for guardrails for build-time compatibility checks, and nothing in the feature set for .NET Core 2.0.

As a result we're removing most of the packages that had this jagged implementation.  The APIs that those packages represented can still be used on the platforms that support them, but we'll no longer pretend that they're portable.  Folks can still use the mscorlib-based portability story (run desktop on NS2.0 compatible frameworks) but it'll have the same characteristics that are present today on a platform like UWP that doesn't have the types.

On a case-by-case basis we can look at bringing back packages with "platform not supported" assemblies.  This isn't great but it can provide a portable surface area with the ability to handle lack of runtime support without doing any dances to avoid a FileNotFound from JIT or dangling ref in AOT.  We'll also consider API additions for feature flags (instead of merely exceptions) to detect support.

If you have some specific asks for packages that you'd like to see remain with a "platform-not-supported" assembly for net standard let me know.  I'm using https://github.com/dotnet/corefx/issues/7528 to track ensuring this is true for packages that are produced today, but many (including ref.emit, runtime.loader) have been disabled.

**Comment by gulbanana** [comment]
Sounds like a good improvement for now. We can always multitarget when something isn't supported in standard - in that respect having the packages gone is going to be nicer than having them present with PlatformNotSupported was.
